### PR TITLE
Support inversion of header matcher logic

### PR DESCRIPTION
**Pull request by marcogschmidt**
We currently support matching a request when a certain header is present or when the value of a specific header matches a given value or a regex (see the relevant [documentation](https://gloo.solo.io/gloo_routing/virtual_services/routes/matching_rules/header_matching/)).

This PR add a new `invert_match` attribute to the [HeaderMatcher](https://gloo.solo.io/api/github.com/solo-io/gloo/projects/gloo/api/v1/proxy.proto.sk/#headermatcher) object that allows to match on a request whose headers do NOT match the specified criteria.
BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/1424

### CODE REVIEW

File: projects/gloo/api/v1/proxy.proto [hunk 1]
```diff
@@ -232,12 +232,8 @@ message Matcher {
     repeated string methods = 8;
 }
 
-//   Internally, Gloo always uses the HTTP/2 *:authority* header to represent the HTTP/1 *Host*
-//   header. Thus, if attempting to match on *Host*, match on *:authority* instead.
-//
-//   In the absence of any header match specifier, match will default to `present_match`
-//   i.e, a request that has the `name` header will match, regardless of the header's
-//   value.
```
Comment (kdorosh) [inline projects/gloo/api/v1/proxy.proto]: why remove this?

File: projects/gloo/pkg/translator/route_config.go [hunk 2]
```diff
@@ -460,10 +460,6 @@ func envoyHeaderMatcher(in []*v1.HeaderMatcher) []*envoyroute.HeaderMatcher {
 				PresentMatch: true,
 			}
 		} else {
-
-			envoyMatch.HeaderMatchSpecifier = &envoyroute.HeaderMatcher_ExactMatch{
-				ExactMatch: matcher.Value,
-			}
```
Comment (kdorosh) [inline projects/gloo/pkg/translator/route_config.go]: again, why remove this? seems like this should be a breaking change

Comment (marcogschmidt) [reply projects/gloo/pkg/translator/route_config.go]: Duplicate logic. See the `else` block just below.

File: projects/gloo/api/v1/proxy.proto (hunk 1 above)
Comment (marcogschmidt) [reply projects/gloo/api/v1/proxy.proto]: This was copy-pasted directly from the [Envoy docs](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/route/route.proto.html?highlight=matcher#route-headermatcher) and does not apply to our API (we do not have a `present_match` specifier, in fact we don't expose any [specifiers](https://github.com/envoyproxy/envoy/blob/373de8b156d873107c4b018b2b1acb07fa51c6d0/api/envoy/api/v2/route/route.proto#L1299) at all).

Comment (marcogschmidt) [reply projects/gloo/api/v1/proxy.proto]: Plus we document the default behavior of `value` directly on the field.

**Comment by marcogschmidt** [comment]
Weird knative test flake, trying again

<details><summary>Error from CI logs</summary>
<p>

```
Step #11 - "test":   /workspace/gopath/src/github.com/solo-io/gloo/projects/clusteringress/pkg/status/status_syncer_test.go:38
Step #11 - "test": [AfterEach] StatusSyncer
Step #11 - "test":   /workspace/gopath/src/github.com/solo-io/gloo/projects/clusteringress/pkg/status/status_syncer_test.go:53
Step #11 - "test": Error from server (Forbidden): namespaces "2b8uykim" is forbidden: User "907026730415@cloudbuild.gserviceaccount.com" cannot delete resource "namespaces" in API group "" in the namespace "2b8uykim"
Step #11 - "test": 
Step #11 - "test": ------------------------------
Step #11 - "test": • Failure in Spec Setup (BeforeEach) [2.636 seconds]
Step #11 - "test": StatusSyncer [BeforeEach] updates kube ingresses with endpoints from the service 
Step #11 - "test": /workspace/gopath/src/github.com/solo-io/gloo/projects/clusteringress/pkg/status/status_syncer_test.go:57
Step #11 - "test": 
Step #11 - "test":   Unexpected error:
Step #11 - "test":       <*errors.StatusError | 0xc00003c750>: {
Step #11 - "test":           ErrStatus: {
Step #11 - "test":               TypeMeta: {Kind: "", APIVersion: ""},
Step #11 - "test":               ListMeta: {SelfLink: "", ResourceVersion: "", Continue: ""},
Step #11 - "test":               Status: "Failure",
Step #11 - "test":               Message: "namespaces is forbidden: User \"907026730415@cloudbuild.gserviceaccount.com\" cannot create resource \"namespaces\" in API group \"\" at the cluster scope",
Step #11 - "test":               Reason: "Forbidden",
Step #11 - "test":               Details: {Name: "", Group: "", Kind: "namespaces", UID: "", Causes: nil, RetryAfterSeconds: 0},
Step #11 - "test":               Code: 403,
Step #11 - "test":           },
Step #11 - "test":       }
Step #11 - "test":       namespaces is forbidden: User "907026730415@cloudbuild.gserviceaccount.com" cannot create resource "namespaces" in API group "" at the cluster scope
Step #11 - "test":   occurred
Step #11 - "test": 
Step #11 - "test":   /workspace/gopath/src/github.com/solo-io/gloo/projects/clusteringress/pkg/status/status_syncer_test.go:51
Step #11 - "test": 
Step #11 - "test":   Full Stack Trace
Step #11 - "test":   	/workspace/gopath/src/github.com/solo-io/gloo/vendor/github.com/onsi/gomega/internal/assertion/assertion.go:75 +0x2e2
Step #11 - "test":   github.com/solo-io/gloo/vendor/github.com/onsi/gomega/internal/assertion.(*Assertion).NotTo(0xc000345080, 0x254b700, 0x3872678, 0x0, 0x0, 0x0, 0x1)
Step #11 - "test":   	/workspace/gopath/src/github.com/solo-io/gloo/vendor/github.com/onsi/gomega/internal/assertion/assertion.go:48 +0x11c
Step #11 - "test":   github.com/solo-io/gloo/projects/clusteringress/pkg/status_test.glob..func1.1()
Step #11 - "test":   	/workspace/gopath/src/github.com/solo-io/gloo/projects/clusteringress/pkg/status/status_syncer_test.go:51 +0x708
Step #11 - "test":   github.com/solo-io/gloo/vendor/github.com/onsi/ginkgo/internal/leafnodes.(*runner).runSync(0xc0000e9920, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
Step #11 - "test":   	/workspace/gopath/src/github.com/solo-io/gloo/projects/clusteringress/pkg/status/status_syncer_suite_test.go:12 +0xf9
Step #11 - "test":   testing.tRunner(0xc00004c700, 0x2306528)
Step #11 - "test":   	/usr/local/go/src/testing/testing.go:865 +0x164
Step #11 - "test":   created by testing.(*T).Run
Step #11 - "test":   	/usr/local/go/src/testing/testing.go:916 +0x69a
Step #11 - "test":   
Step #11 - "test": ------------------------------
Step #11 - "test": 
Step #11 - "test": 
Step #11 - "test": Summarizing 1 Failure:
Step #11 - "test": 
Step #11 - "test": [Fail] StatusSyncer [BeforeEach] updates kube ingresses with endpoints from the service 
Step #11 - "test": /workspace/gopath/src/github.com/solo-io/gloo/projects/clusteringress/pkg/status/status_syncer_test.go:51
Step #11 - "test": 
Step #11 - "test": Ran 1 of 1 Specs in 2.640 seconds
Step #11 - "test": FAIL! -- 0 Passed | 1 Failed | 0 Pending | 0 Skipped
Step #11 - "test": --- FAIL: TestStatusSyncer (2.64s)
Step #11 - "test": FAIL
```
</p>
</details>

/kick

**Comment by marcogschmidt** [comment]
Test failure was caused by permissions on CI account. Should be fixed now.

/kick

**Pull request merged**